### PR TITLE
Fix for python gemm generation

### PR DESCRIPTION
**Pull request by vidyasiv**
## Description

- We were not clearing out the tile_descriptions in between the math instructions loops leading to incorrect generation. See internal ticket for more details on issue.
- Added testing for generation and custom tile shapes. 

Sample code that tests this
```
import torch
from torch.utils import _pytree as pytree
from torch._inductor.codegen.cutlass.utils import _gen_ops_cached
from torch._inductor.codegen.cutlass.utils import try_import_cutlass
assert try_import_cutlass()
import cutlass_library


# arch=12 means BMG.
# 20250302 means oneapi version, may not be used.
full_ops = _gen_ops_cached("12", "20250302", "xpu")
ops = pytree.tree_flatten(full_ops)[0]
f16_cnt = 0
bf16_cnt = 0
names = []
for op in ops:
    if op.A.element == cutlass_library.library.DataType.f16:
        f16_cnt += 1
    elif op.A.element == cutlass_library.library.DataType.bf16:
        bf16_cnt += 1
print("f16 count: ", f16_cnt)
print("bf16 count: ", bf16_cnt)
```

Error output:
```
f16 count:  140
bf16 count:  340
```

After the fix output is
```
f16 count:  120
bf16 count:  120
```

## Type
- [x] Bug  - [ ] Feature  - [ ] Performance  - [ ] Refactor

## Testing
- [x] Tests pass  - [ ] Xe12  - [ ] Xe20

## Performance
| Metric | Before | After |
|--------|--------|-------|
|        |        |       |

## References
Fixes #

## Checklist
- [x] Copyright  
- [ ] Co-pilot Review  
- [x] Deprecated APIs not used

### CODE REVIEW

**Review by jiyang1011 — APPROVED** [review]
LGTM

**Review by taozha2 — APPROVED** [review]
